File: LeetCode/src/array/FindPivotIndex.java
package array;

public class FindPivotIndex {
    public int pivotIndex(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        prefix[0] = nums[0];

        for (int i = 1; i < n; i++) {
            prefix[i] = prefix[i - 1] + nums[i];
        }

        int left = 0, right = 0;
        for (int i = 0; i < n; i++) {
            left = prefix[i] - nums[i];
            right = prefix[n - 1] - prefix[i];

            if (left == right) {
                return i;
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        FindPivotIndex soltuion = new FindPivotIndex();
        int[] nums1 = {1, 7, 3, 6, 5, 6};
        System.out.println(soltuion.pivotIndex(nums1)); // Output: 3

        int[] nums2 = {1, 2, 3};
        System.out.println(soltuion.pivotIndex(nums2)); // Output: -1

        int[] nums3 = {2, 1, -1};
        System.out.println(soltuion.pivotIndex(nums3)); // Output: 0
    }
}

File: LeetCode_Python/prefixsum/FindPivotIndex.py
class FindPivotIndex:
    def pivotIndex(self, nums: list[int]) -> int:
        s = sum(nums)
        cur = 0

        for i, n in enumerate(nums):
            if cur == s - cur - n:
                return i
            cur += n

        return -1

    # Time Complexity: O(n), where n is the length of nums
    # Space Complexity: O(1)

    def pivotIndexByArray(self, nums: list[int]) -> int:
        n = len(nums)
        left = [0] * n
        right = [0] * n
        left[0] = nums[0]
        right[n - 1] = nums[n - 1]

        for i in range(1, n):
            left[i] = left[i - 1] + nums[i]

        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] + nums[i]

        for i in range(n):
            if left[i] == right[i]:
                return i

        return -1

    # Time Complexity: O(n), where n is the length of nums
    # Space Complexity: O(n), where n is the length of nums


if __name__ == "__main__":
    sol = FindPivotIndex()

    print("Expected Output: 3")
    print("Actual Output:", sol.pivotIndex([1, 7, 3, 6, 5, 6]))
    print("Actual Output:", sol.pivotIndexByArray([1, 7, 3, 6, 5, 6]))

    print("Expected Output: -1")
    print("Actual Output:", sol.pivotIndex([1, 2, 3]))
    print("Actual Output:", sol.pivotIndexByArray([1, 2, 3]))

    print("Expected Output: 0")
    print("Actual Output:", sol.pivotIndex([2, 1, -1]))
    print("Actual Output:", sol.pivotIndexByArray([2, 1, -1]))
